internal/controllers/utils/events.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"sort"
+	"time"
+)
+
+const (
+	// Generic Event states
+	// InProgress indicates a in progress state
+	InProgress = "InProgress"
+	// InProgress indicates a completed state
+	Completed = "Completed"
+)
+
+// Event is the generic representation of an event
+type Event struct {
+	ObjectID  string
+	Timestamp time.Time
+	State     string
+}
+
+// EventHistory contains a slice of events
+type EventHistory struct {
+	History []*Event
+}
+
+// IsTimedOut proceses a process history an determines the overall state implied
+// by the events.
+// returns true if for any ObjectID:
+// - the overall state was InProgress from the initial time for more than timeout
+// - the overall state was InProgress from the last transition to InProgress for more than timeout
+// returns false if for any ObjectID:
+// - the last transition is Completed
+// - no events to process
+func (h EventHistory) IsTimedOut(now time.Time, timeout time.Duration) bool {
+	// Sort events chronologically from old to new
+	sort.Slice(h.History, func(i, j int) bool {
+		return h.History[i].Timestamp.Before(h.History[j].Timestamp)
+	})
+
+	// currentState stores the running value of the State (InProgress or Completed) for each ObjectID
+	currentState := map[string]string{}
+	// Initialize the current state to NonCompliant
+	for _, event := range h.History {
+		currentState[event.ObjectID] = InProgress
+	}
+	// initialTime records the initial Event for all ObjectIDs
+	initialTime := time.Time{}
+	// resetTime is a recording a zero time
+	resetTime := time.Time{}
+	// lastCompleted records the last time the overall state for all event transitioned to Completed
+	lastCompleted := time.Time{}
+	// lastInProgress records the last time the overall state for all event transitioned to InProgress
+	lastInProgress := time.Time{}
+	// Process each event chronologically one by one and recalculate the
+	// current state (InProgress or Completed)
+	for _, event := range h.History {
```

Comment:
It's not clear to me why we cannot rely on the policy controller to have computed the correct state and it has to be recreated. I understand it's trying to follow the k8s principles and relying just on observed data, but I think in this operator, we can safely use what we are observing in the status of other CRs that are not managed by the current operator.
I think we only care if the Policy is currently NonCompliant/Pending and to see since when.
cc @pixelsoccupied

Reply:
The advantage of processing the history is to detect the times in history when the policies where all compliant, even if these are not the latest. Since these times reset the timeout timer. So in the example below, if the timeout is set at 6s, at the 6s time, we would time out if we look only at the current state (non compliant). By looking at the history, we can see that we were compliant before at the 5s time so the timer was reset, counting from zero again. So the next timeout would be at time 11s :
![image](https://github.com/user-attachments/assets/f271f816-a739-4304-b1df-33483f8b47be)
@irinamihai did I understand the issue raised in this diagram properly? or this is not the problem described?

Reply:
I think what Irina is saying is that we can simply read the policy CR directly...instead of doing a conversion to something custom (EventHistory)? Basically we really need `BuildPoliciesEventHistory` func?

Reply:
I created EventHistory on purpose to de-correlate event processing from the specific k8s policies event history. It makes it easier to test and understand without mixing k8s implementation details with it. Also, if needed we could reuse the code if there is another similar object with events that we need to process a timeout from. The BuildPoliciesEventHistory is just the specific part of this process, it just very specific to policies events and convert this specific flavor of events into something generic that we do not have to retest. If we get the timeout for another type of historical events, this build function would be the only one needed.